db/db_impl.cc

```diff
@@ -1501,6 +1503,14 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& log_numbers,
       WriteBatchInternal::SetContents(&batch, record);
       SequenceNumber sequence = WriteBatchInternal::Sequence(&batch);
 
+      // If we detect a gap in our log stream then we have been reading in
+      // stale log data and should remove all recovered 2pc transactions because
+      // the commit/rollback marker could have been in the missing log.
+      if (first_record_in_log && !first_log_in_recover &&
+          (*next_sequence != sequence)) {
+        DeleteAllRecoveredTransactions();
+      }
```

Comment:
Keep in mind that a hole of sequence IDs can happen if there are some writes are written using disableWAL=true in the middle. So this can fail many existing services. It will be safer to just do it if 2pc=true, and explain in the comments.